src/detectors/BTOF/BTOF.cc

```diff
@@ -55,53 +50,75 @@ void InitPlugin(JApplication *app) {
         app
     ));         // Hit reco default config for factories
 
-    int BarrelTOF_ID = 0;
-    try {
-        auto detector = app->GetService<DD4hep_service>()->detector();
-        BarrelTOF_ID = detector->constant<int>("BarrelTOF_ID");
-    } catch(const std::runtime_error&) {
-        // Nothing
-    }
-    PIDLookupConfig pid_cfg {
-      .filename="calibrations/tof.lut",
-      .system=BarrelTOF_ID,
-      .pdg_values={11, 211, 321, 2212},
-      .charge_values={1},
-      .momentum_edges={0.0, 0.3, 0.6, 0.9, 1.2, 1.5, 1.8, 2.1, 2.4, 2.7, 3.0, 3.3, 3.6, 3.9, 4.2, 4.5, 4.8, 5.1, 5.4, 5.7, 6.0},
-      .polar_edges={2.50, 10.95, 19.40, 27.85, 36.30, 44.75, 53.20, 61.65, 70.10, 78.55, 87.00, 95.45, 103.90, 112.35, 120.80, 129.25, 137.70, 146.15, 154.60},
-      .azimuthal_binning={0., 360., 360.}, // lower, upper, step
-      .momentum_bin_centers_in_lut=true,
-      .polar_bin_centers_in_lut=true,
-    };
+  app->Add(new JOmniFactoryGeneratorT<BTOFChargeSharing_factory>(
+      "BTOFChargeSharing",
+      {"TOFBarrelHits"},
+      {"TOFBarrelSharedHits"},
+      {},
+      app
+  ));
+
+  app->Add(new JOmniFactoryGeneratorT<TOFPulseGeneration_factory>(
+      "BTOFPulseGeneration",
+      {"TOFBarrelSharedHits"},
+      {"TOFBarrelPulse"},
+      {},
+      app
+  ));
+
+  app->Add(new JOmniFactoryGeneratorT<TOFPulseDigitization_factory>(
+      "BTOFPulseDigitization",
+      {"TOFBarrelPulse"},
+      {"TOFBarrelADCTDC"},
```

Comment:
Can we get this code running by default? What stops us from using this in the tracking?

Reply:
To have the code running, we need to add "TOFBarrelADCTDC" to src/services/io/podio/JEventProcessorPODIO.cc.

Reply:
Currently, TOFBarrelADCTDC is not used anywhere downstream so I leave it out. You can still call it with "eicrecon -Ppodio:output_collections=TOFBarrelADCTDC output.edm4hep.root" etc. My plan is to push the clustering algorithm in my future pull request where it calculate hit location and time from TOFBarrelADCTDC. The reconstructed location will replace the original TOFBarrelRecHits so it will be used by the tracking algorithm downstream. I don't think we should write TOFBarrelADCTDC to output by default because saving individual pulses take up too (relatively) large amount of disk space, but if you believe it's a good idea to put it in JEventProcessorPODIO.cc, then I don't think that will be too much of an issue either.

Reply:
I thought even with sharing digitization would output only a handful of RawTrackerHits. But yeah, if it's a large output, we can't have that saved, unfortunately.

Okay, how do we continuously test this then? Can you provide some unit tests like those in `src/tests/algorithms_test`?

Reply:
Okay sorry I remembered incorrectly. "TOFBarrelPulse" saves the entire pulse, but "TOFBarrelADCTDC" only saves the ADC and TDC values. There is charge spreading so there could be 3-6 ADC-TDC pairs for each BTOF hits, but it is still just a handful of numbers. I take that back, I think we can save TOFBarrelADCTDC, but not TOFBarrelPulse.

We can still test it, like see if TDC vs geant hit time follows some rough linear relationship. We can also try to estimate if the hit position vs ADC weighted position follows linear relationship. 

Are src/tests/algorithms_test tests being run automatically everytime a pull request is made? I think I can come up with some sort of linearity test.

Reply:
> Are src/tests/algorithms_test tests being run automatically everytime a pull request is made?

Yes

> I think I can come up with some sort of linearity test.

That would be great. Note that those tests intended to check individual components (which is sort of useful on its own). If you want to setup a more complete chain, it's a bit more involved, but there are some examples in `src/tests/omnifactory_test` (you will need to setup a separate subdirectory in `src/tests` for that).

Reply:
I tried to write a simple test for BTOFChargeSharing, but I can't get cellIDPositionConverter to work. I tried to put mock TPC hit in algorithmsInit.cc, but it seems like even if readout and segmentation are set correctly, there is no physical coordinates that corresponds to those mock hits. BTOFChargeSharing can't run because it needs to convert global coordinates to local for charge sharing calculation. I am sure cellIDPositionConverter is responsible for my error because the always errors out when I ask for cell position from cellIDPositionConverter.

I think TOFPulseGeneration and TOFPulseDigitization don't need cellIDPositionConverter and I can still test those two classes. Do you have any suggestions regarding BTOFChargeSharing? Thanks for any of your suggestions.

Reply:
What you could do for the charge sharing test is refactor the code so that there is a core function which takes the positions as arguments, having converted from CellID beforehand. This function would probably best be in place of your loop over neighbours.

Reply:
Yes, there is. The "mock" detectors are setup here:
https://github.com/eic/EICrecon/blob/45eddf45b5e426435dd8de7467a0068629d7d638/src/tests/algorithms_test/algorithmsInit.cc#L31-L45
The cellID spec should be obvious how to modify, but let me know if you run into trouble.
I believe you can also add some fake volumes to make `cellIDPositionConverter` work, but I have not done so myself.

If you have some working tests please push them to the branch. If there are tests that you've implemented that should work, you can push too, we can debug here together.